### ISSUE TITLE
[1.3.0-RC2] credentials are not cached across the subprojects

### ISSUE THREAD

**Issue by BillyAutrey** [sbt/sbt#4802]
## steps
sbt version: 1.3.0-RC2

## problem
When trying to compile code with Lightbend Commercial credentials, coursier fails to resolve those dependencies.

```
[error] Error downloading com.lightbend.cinnamon:cinnamon-slf4j-events:2.11.1
[error]   Not found
[error]   not found: <userhome>/.ivy2/local/com.lightbend.cinnamon/cinnamon-slf4j-events/2.11.1/ivys/ivy.xml
[error]   not found: https://repo1.maven.org/maven2/com/lightbend/cinnamon/cinnamon-slf4j-events/2.11.1/cinnamon-slf4j-events-2.11.1.pom
[error]   unauthorized: https://repo.lightbend.com/commercial-releases/com/lightbend/cinnamon/cinnamon-slf4j-events/2.11.1/cinnamon-slf4j-events-2.11.1.pom (Bintray)
```

## expectation
Compilation should complete as normal, with credentials found in the location indicated in .sbt files.

## notes
Repository configuration:

> project/cinnamon.sbt
```
addSbtPlugin("com.lightbend.cinnamon" % "sbt-cinnamon" % "2.11.1")
// Credentials and resolver to download the Cinnamon Telemetry libraries
credentials += Credentials(Path.userHome / ".lightbend" / "commercial.credentials")
resolvers += Resolver.url("lightbend-commercial", url("https://repo.lightbend.com/commercial-releases"))(Resolver.ivyStylePatterns)
```

If I set `ThisBuild / useCoursier := true` with 1.3.0-RC2, the build will fail.
If I set `ThisBuild / useCoursier := false` with 1.3.0-RC2, the build will succeed.

Setting `host = repo.lightbend.com` in my `commercial.credentials` file does not resolve the issue.  Nor does leaving it as `host = dl.bintray.com`.

Downgrading to 1.2.7 also allows the build to succeed.

**Comment by BillyAutrey** [comment]
Possibly related - https://github.com/coursier/coursier/issues/1177

**Comment by dejan2609** [comment]
Same/similar here. 

Environment:
* Nexus Repository Manager OSS 2.14.13-01
* sbt **1.2.8** -->> build works as expected ✔
* sbt **1.3.0-RC2** build fails ❌ (see error stack trace below)
* sbt **1.3.0-RC2** with `ThisBuild / useCoursier := false`  -->> build works as expected ✔
	
build.sbt:
	
```
resolvers ++= {
  Seq(
     Resolver.mavenLocal,
    "Our Nexus repository" at "https://.../nexus/content/groups/public"
  )
}
credentials += Credentials("Sonatype Nexus Repository Manager", "abc.xyz.com", "username", "password")
```

```
[error] Error downloading some.dependency.name
[error]   Not found
[error]   not found: C:\Users\....\ivys\ivy.xml
[error]   not found: https://repo1.maven.org/maven2/..........
[error]   unauthorized: https://..../nexus/content/groups/public/some.dependency.name.pom
[error]         at lmcoursier.CoursierDependencyResolution.unresolvedWarningOrThrow(CoursierDependencyResolution.scala:225)
[error]         at lmcoursier.CoursierDependencyResolution.$anonfun$update$30(CoursierDependencyResolution.scala:194)
[error]         at scala.util.Either$LeftProjection.map(Either.scala:573)
[error]         at lmcoursier.CoursierDependencyResolution.update(CoursierDependencyResolution.scala:194)
[error]         at sbt.librarymanagement.DependencyResolution.update(DependencyResolution.scala:56)
[error]         at sbt.internal.LibraryManagement$.resolve$1(LibraryManagement.scala:47)
[error]         at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$12(LibraryManagement.scala:97)
[error]         at sbt.util.Tracked$.$anonfun$lastOutput$1(Tracked.scala:69)
[error]         at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$20(LibraryManagement.scala:110)
[error]         at scala.util.control.Exception$Catch.apply(Exception.scala:228)
[error]         at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$11(LibraryManagement.scala:110)
[error]         at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$11$adapted(LibraryManagement.scala:91)
[error]         at sbt.util.Tracked$.$anonfun$inputChanged$1(Tracked.scala:150)
[error]         at sbt.internal.LibraryManagement$.cachedUpdate(LibraryManagement.scala:124)
[error]         at sbt.Classpaths$.$anonfun$updateTask0$5(Defaults.scala:2895)
[error]         at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error]         at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error]         at sbt.std.Transform$$anon$4.work(Transform.scala:67)
[error]         at sbt.Execute.$anonfun$submit$2(Execute.scala:280)
[error]         at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:19)
[error]         at sbt.Execute.work(Execute.scala:289)
[error]         at sbt.Execute.$anonfun$submit$1(Execute.scala:280)
[error]         at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:178)
[error]         at sbt.CompletionService$$anon$2.call(CompletionService.scala:37)
[error]         at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error]         at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error]         at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error]         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error]         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error]         at java.lang.Thread.run(Thread.java:748)
```

**Comment by dejan2609** [comment]
Also, this would be just awesome: https://github.com/coursier/coursier/issues/1160

**Comment by eed3si9n** [comment]
I was not able to reproduce this bug.

I wiped out `~/.coursier/cache/v1/https/repo.lightbend.com/commercial-releases/` and run a build:

```
[info] Loading global plugins from /Users/eed3si9n/.sbt/1.0/plugins
[info] Loading settings for project cinnamon-sample-build from plugins.sbt ...
[info] Loading project definition from /Users/eed3si9n/work/quicktest/cinnamon-sample/project
[info] Updating
https://repo.lightbend.com/commercial-releases/com.lightbend.cinnamon/sbt-cinnamon/scala_2.12/sbt_1.0/2.11.1/ivys/ivy.xml
  100.0% [##########] 2.7 KiB (4.1 KiB / s)
[info] Resolved  dependencies
[info] Fetching artifacts of
https://repo.lightbend.com/commercial-releases/com.lightbend.cinnamon/sbt-cinnamon/scala_2.12/sbt_1.0/2.11.1/jars/sbt-cinnamon.jar
  100.0% [##########] 24.3 KiB (32.3 KiB / s)
[info] Fetched artifacts of
[info] Loading settings for project root from build.sbt ...
[info] Set current project to cinnamon-sample (in build file:/Users/eed3si9n/work/quicktest/cinnamon-sample/)
[info] Welcome to sbt 1.3.0-RC2.
```

One thing I did notice was that my file was called **~/.lightbend/commercial.credential** (singular) and I had to rename it to **~/.lightbend/commercial.credentials** to match your test and current documentation. I wonder if this was changed at some point.

**Comment by eed3si9n** [comment]
It doesn't look like there's 1.3.0-RC2 specific issue.

**Comment by BillyAutrey** [comment]
Some additional context in case anyone Googles for this issue later, and ends up here.

* My project was a multiple module build.
* Adding `ThisBuild / credentials += Credentials(Path.userHome / ".lightbend" / "commercial.credentials")` to my `build.sbt` allowed coursier to properly authenticate for the artifacts.
* Resolvers propagate down to all submodules, from the plugin.  No action required for resolvers specifically.

**Comment by eed3si9n** [comment]
At some point I've seen this, but it has slipped my mind.

https://github.com/sbt/sbt/blob/9fa25de84cf6fb521aa5f161c866837200ec17c4/main/src/main/scala/sbt/Defaults.scala#L2271-L2275

https://github.com/sbt/librarymanagement/blob/c236d111b833ca6d048cdadfb669317eeee17b34/ivy/src/main/scala/sbt/librarymanagement/ivy/Credentials.scala#L58-L62

https://github.com/sbt/librarymanagement/blob/c236d111b833ca6d048cdadfb669317eeee17b34/ivy/src/main/scala/sbt/librarymanagement/ivy/Credentials.scala#L19-L21

```scala
  /** Add the provided credentials to Ivy's credentials cache.*/
  def add(realm: String, host: String, userName: String, passwd: String): Unit =
    CredentialsStore.INSTANCE.addCredentials(realm, host, userName, passwd)
```

Basically for Ivy (sbt's use of Ivy), there's a global singleton object that accumulates the credential information across the build. This breaks away from normal sbt's scoping behavior, but it might be necessary to mimic this for lm-coursier as well for backward compatibility.

**Comment by francisdb** [comment]
Is there a way to show/log what credentials sbt is using to go to nexus?

**Comment by eed3si9n** [comment]
@francisdb Not atm. I might create a task that can.

**Comment by francisdb** [comment]
I just tried `1.3.0-RC3` and still fails to look up our private plugin (unauthorized on nexus repo)
credentials registered in `~/.sbt/1.0/global.sbt`
Should I do anything to get this working?

Disable the plugin and tried resolving normal dependencies (multiproject build) but also fails with unauthorized.
`show credentials` and `show allCredentials` both contain my credentials
Somehow coursier still is not using those.

**Comment by eed3si9n** [comment]
Let me validate using the repro steps provided in this issue first.

**Comment by sbrunk** [comment]
Just tested with our private nexus and `1.3.0-RC3` and and I'm seeing unauthorized errors like @francisdb.

**Comment by rory-ye-nv** [comment]
same problem here, change the property file as coursier doesn't work.

**Comment by 2m** [comment]
I think I am hitting this bug, but in a different scenario. After switching from 1.2.8 to 1.3.0-RC3, projects do not load correctly with [sbt-authors](https://github.com/2m/authors) in the build. This plugin requires an additional resolver to load all of its dependencies, and it seems that additional resolver is not taken into consideration by sbt.

I have the following in `~/.sbt/1.0/plugins/build.sbt`:

```
addSbtPlugin("lt.dvim.authors" % "sbt-authors" % "1.0.2")
resolvers += Resolver.bintrayRepo("jypma", "maven") // for ts-reaktive
```

And then if a project uses sbt `1.3.0-RC3` it fails to load with:

```
[error]   not found: https://repo1.maven.org/maven2/com/tradeshift/ts-reaktive-marshal-akka_2.12/0.11.0/ts-reaktive-marshal-akka_2.12-0.11.0.pom
[error]   not found: /home/martynas/.ivy2/local/com.tradeshift/ts-reaktive-marshal-akka_2.12/0.11.0/ivys/ivy.xml
[error]   not found: https://repo.scala-sbt.org/scalasbt/sbt-plugin-releases/com.tradeshift/ts-reaktive-marshal-akka_2.12/0.11.0/ivys/ivy.xml
[error]   not found: https://repo.typesafe.com/typesafe/ivy-releases/com.tradeshift/ts-reaktive-marshal-akka_2.12/0.11.0/ivys/ivy.xml
```

**Comment by eed3si9n** [comment]
I confirmed that the original issue that was reported about Cinnamon looks like it's fixed using RC-3. I added the following to `~/.sbt/1.0/credentials.sbt`:

```scala
credentials += Credentials(Path.userHome / ".lightbend" / "something")
```

**Comment by eed3si9n** [comment]
@2m Try adding the resolver to `~/.sbt/1.0/global.sbt`.

**Comment by 2m** [comment]
That did not help. Should I open another ticket?

**Comment by eed3si9n** [comment]
@2m I think so.

**Comment by limansky** [comment]
@eed3si9n it seems like this issue is still reproducible with 1.3.0-RC5.

**Comment by kyleu** [comment]
Reproducible with 1.3.0 final as well

**Comment by eed3si9n** [comment]
The problem described in this issue I think has been resolved. If there's some other credential related issue that's not https://github.com/sbt/sbt/issues/4852, please open a separate issue describing the repro steps.
